core/src/main/java/org/apache/iceberg/avro/BuildAvroProjection.java

```diff
@@ -82,28 +82,53 @@ public Schema record(Schema record, List<String> names, Iterable<Schema.Field> s
     List<Schema.Field> updatedFields = Lists.newArrayListWithExpectedSize(struct.fields().size());
     List<Types.NestedField> expectedFields = struct.fields();
     for (int i = 0; i < expectedFields.size(); i += 1) {
-      Types.NestedField field = expectedFields.get(i);
-
+      Types.NestedField expectedField = expectedFields.get(i);
       // detect reordering
-      if (i < fields.size() && !field.name().equals(fields.get(i).name())) {
+      if (i < fields.size() && !expectedField.name().equals(fields.get(i).name())) {
         hasChange = true;
       }
 
-      Schema.Field avroField = updateMap.get(AvroSchemaUtil.makeCompatibleName(field.name()));
+      Schema.Field avroField = updateMap.get(AvroSchemaUtil.makeCompatibleName(expectedField.name()));
 
       if (avroField != null) {
-        updatedFields.add(avroField);
-
+        // if the expectedField has a defaultValue, but the avroField does not, we need to
+        // create a newField to copy over the non-null default value
+        if (expectedField.hasDefaultValue() && !AvroSchemaUtil.hasNonNullDefaultValue(avroField)) {
+          Schema newFiledSchema = (expectedField.isOptional()) ?
+              AvroSchemaUtil.toOption(AvroSchemaUtil.convert(expectedField.type()), true) :
+              AvroSchemaUtil.convert(expectedField.type());
+          Schema.Field newField =
+              new Schema.Field(avroField.name(), newFiledSchema, avroField.doc(), expectedField.getDefaultValue());
+          newField.addProp(AvroSchemaUtil.FIELD_ID_PROP, expectedField.fieldId());
+          updatedFields.add(newField);
+        } else {
+          // otherwise (i.e., expectedFiled has no default value, or it is null) we can use avroField as is
+          updatedFields.add(avroField);
+        }
       } else {
-        Preconditions.checkArgument(
-            field.isOptional() || MetadataColumns.metadataFieldIds().contains(field.fieldId()),
-            "Missing required field: %s", field.name());
-        // Create a field that will be defaulted to null. We assign a unique suffix to the field
-        // to make sure that even if records in the file have the field it is not projected.
-        Schema.Field newField = new Schema.Field(
-            field.name() + "_r" + field.fieldId(),
-            AvroSchemaUtil.toOption(AvroSchemaUtil.convert(field.type())), null, JsonProperties.NULL_VALUE);
-        newField.addProp(AvroSchemaUtil.FIELD_ID_PROP, field.fieldId());
+        // here the expectedField is missing from the file schema, so we verify it is either
+        // an optional field, a metadata column or one that has default value
+        Preconditions.checkArgument(expectedField.isOptional() ||
+                expectedField.fieldId() == MetadataColumns.ROW_POSITION.fieldId() || expectedField.hasDefaultValue(),
```

Comment:
I think we should keep the previous check `MetadataColumns.metadataFieldIds()` not just `MetadataColumns.ROW_POSITION`.

Reply:
Good catch @rzhang10 ! This line in Iceberg 0.9 is using the `MetadataColumns.ROW_POSITION` and I missed the change while cherry picking merge! Thanks a lot!